### PR TITLE
chore: force config set remove cli dependency

### PR DESCRIPTION
**Pull request by RitamAgrawal**
### What does this PR do?
- Removes CLI dependency in ForceConfigSet
- Incorporates sfdx-core V3
### Functionality Before
Implement `force:config:set` command using _SfdxCommandLetExecutor_, making it dependent on CLI

### Functionality After
Implement `force:config:set` using _LibraryCommandLetExecutor_, removing the dependency from CLI. It is using sfdx-core lib.

### CODE REVIEW

File: packages/salesforcedx-utils-vscode/src/config/configUtil.ts [hunk 1]
```diff
@@ -134,4 +136,20 @@ export class ConfigUtil {
       : undefined;
     return username ? String(username) : undefined;
   }
+
+  public static async setDefaultUsernameOrAlias(usernameOrAlias: string): Promise<void> {
+    const originalDirectory = process.cwd();
```
Comment (RitamAgrawal) [inline packages/salesforcedx-utils-vscode/src/config/configUtil.ts:141]: Need to add test for this one.

File: packages/salesforcedx-vscode-core/test/vscode-integration/commands/forceConfigSet.test.ts [hunk 2]
```diff
@@ -5,35 +5,85 @@
...
+  afterEach(() => {
+    sandbox.restore();
+  });
+
+  it('should set config with the given username or alias', async () => {
+    orgStub.resolves();
+    await forceConfigSet(usernameOrAlias);
+    expect(setDefaultUsernameOrAliasStub.callCount).to.equal(1);
+    expect(setDefaultUsernameOrAliasStub.calledWith(usernameOrAlias)).to.equal(true);
+    // expect(configWriteSpy.callCount).to.equal(1);
```
Comment (RitamAgrawal) [inline packages/salesforcedx-vscode-core/test/vscode-integration/commands/forceConfigSet.test.ts:46]: config.write is called within ConfigUtil.setDefaultUsernameOrAlias function.

File: packages/salesforcedx-vscode-core/test/vscode-integration/commands/forceConfigSet.test.ts [hunk 3]
```diff
@@ -5,35 +5,76 @@
...
+  it('should display correct output to user', async () => {
+    const outputTableRow = { name: DEFAULT_USERNAME_KEY, val: usernameOrAlias, success: String(true) };
+    const forceConfigSetInstance = new ForceConfigSetExecutor(usernameOrAlias);
+    const formatOutput = (forceConfigSetInstance as any).formatOutput(outputTableRow);
+    expect(tableSpy.callCount).to.equal(1);
+    expect(formatOutput).to.contain(nls.localize(CONFIG_SET_NAME), DEFAULT_USERNAME_KEY);
+    expect(formatOutput).to.contain(usernameOrAlias, String(true));
+  });
+
+  it('should not set config with an invalid username or alias', async () => {
```
Comment (klewis-sfdc) [inline packages/salesforcedx-vscode-core/test/vscode-integration/commands/forceConfigSet.test.ts:67]: This test might not be needed.  It looks like its logic is also covered in the next test, so I think it'd be ok to remove this one.

File: packages/salesforcedx-vscode-core/test/vscode-integration/commands/forceConfigSet.test.ts [hunk 4]
```diff
@@ -5,35 +5,76 @@
...
+    expect(channelSpy.callCount).to.equal(2);
+  });
+
+  it('should display error message in output channel', async () => {
+    setDefaultUsernameOrAliasStub.throws(new Error(errorMessage));
+    await forceConfigSet(usernameOrAlias);
+    expect(channelSpy.callCount).to.equal(2);
+    expect(channelSpy.lastCall.args.length).to.equal(1);
+    expect(channelSpy.lastCall.args[0]).to.contain(errorMessage);
+  }).timeout(10000);
```
Comment (klewis-sfdc) [inline packages/salesforcedx-vscode-core/test/vscode-integration/commands/forceConfigSet.test.ts:79]: Is this timeout still needed here?  Does the test fail without it?

File: packages/salesforcedx-utils-vscode/src/config/configUtil.ts [hunk 5]
```diff
@@ -8,10 +8,12 @@
 import {
   Config,
   ConfigAggregator,
+  Org,
   OrgConfigProperties,
   SfConfigProperties,
   StateAggregator
 } from '@salesforce/core';
+import { DEFAULT_USERNAME_KEY, getRootWorkspacePath } from '..';
```
Comment (klewis-sfdc) [inline packages/salesforcedx-utils-vscode/src/config/configUtil.ts:16]: Looks like DEFAULT_USERNAME_KEY is unused and can be removed here.

File: packages/salesforcedx-utils-vscode/src/config/configUtil.ts [hunk 6]
```diff
@@ -134,4 +136,22 @@ export class ConfigUtil {
       : undefined;
     return username ? String(username) : undefined;
   }
+
+  public static async setDefaultUsernameOrAlias(usernameOrAlias: string): Promise<void> {
```
Comment (klewis-sfdc) [inline packages/salesforcedx-utils-vscode/src/config/configUtil.ts:140]: I think configUtil.test.ts needs some tests to cover this new function.  Those tests should cover:

* That the cwd is switched to the project root if necessary (and switched back to the orig cwd after)
* That the core lib is being called correctly.  i.e., when called with <username>, Org.create is called
  with <username>, and then Config.set() is called with "<OrgConfigProperties.TARGET_ORG>, <username>", etc.

File: packages/salesforcedx-vscode-core/test/vscode-integration/commands/forceConfigSet.test.ts [hunk 7]
```diff
@@ -5,35 +5,76 @@
  * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
  */
 
+import { Config, Org } from '@salesforce/core';
```
Comment (klewis-sfdc) [inline packages/salesforcedx-vscode-core/test/vscode-integration/commands/forceConfigSet.test.ts:8]: Looks like Config and Org are unused here and can be removed.

**Review by klewis-sfdc — COMMENTED** [review]
Looking really good Ritam.  The tests for forceConfigSet look great.  I think there should be some tests added for the new function in configUtil.  You can look at the tests for config agg provider for help, and let me know if you'd like to discuss!

File: packages/salesforcedx-vscode-core/test/vscode-integration/commands/forceConfigSet.test.ts (hunk 4 above)
Comment (RitamAgrawal) [reply packages/salesforcedx-vscode-core/test/vscode-integration/commands/forceConfigSet.test.ts]: Tests are passing without the timeout. Removed it.

File: packages/salesforcedx-utils-vscode/src/config/configUtil.ts [hunk 8]
```diff
@@ -144,6 +146,28 @@ export class ConfigUtil {
...
+    const originalDirectory = process.cwd();
+    // In order to correctly setup Config, the process directory needs to be set to the current workspace directory
+    const workspacePath = workspaceUtils.getRootWorkspacePath();
+    try {
+      if (usernameOrAlias) { // check if username is an empty string
+        await Org.create({ aliasOrUsername: usernameOrAlias }); // check if username is valid
+      }
```
Comment (gbockus-sf) [inline packages/salesforcedx-utils-vscode/src/config/configUtil.ts:157]: nit: I could see pulling the above logic out into a `isUsernameValid` function.  It's a little concerning now that it's not very clear what happens if it is invalid.

File: packages/salesforcedx-utils-vscode/test/jest/config/configUtil.test.ts [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+  it('should change the current working directory to the original working directory', async () => {
+    const username = 'vscodeO';
+    await ConfigUtil.setDefaultUsernameOrAlias(username);
+    expect(workspacePathStub).toHaveBeenCalledTimes(1);
+    expect(chdirStub).toHaveBeenCalledTimes(2);
+    expect(chdirStub).toHaveBeenNthCalledWith(1, fakeWorkspace);
+    expect(chdirStub).toHaveBeenNthCalledWith(2, fakeOriginalDirectory);
+  });
+
+  it('should set username or alias as default if it is an empty string', async () => {
```
Comment (gbockus-sf) [inline packages/salesforcedx-utils-vscode/test/jest/config/configUtil.test.ts:44]: is this test name accurate?  It's not default but the passed empty string correct? I might go with, `Should be able to set username to empty string.`

File: packages/salesforcedx-vscode-core/src/commands/forceConfigSet.ts [hunk 10]
```diff
@@ -4,33 +4,72 @@
  * Licensed under the BSD 3-Clause license.
  * For full license text, see LICENSE.txt file in the repo root or https://opensource.org/licenses/BSD-3-Clause
  */
+import { ConfigUtil, LibraryCommandletExecutor } from '@salesforce/salesforcedx-utils-vscode/out/src';
+import { Row, Table } from '@salesforce/salesforcedx-utils-vscode/out/src/output';
+import { ContinueResponse } from '@salesforce/salesforcedx-utils-vscode/src/types';
```
Comment (gbockus-sf) [inline packages/salesforcedx-vscode-core/src/commands/forceConfigSet.ts:9]: The three above imports should be updated to just import from `@salesforce/salesforcedx-utils-vscode`

File: packages/salesforcedx-vscode-core/src/constants.ts [hunk 11]
```diff
@@ -25,6 +25,12 @@ export const TELEMETRY_OPT_OUT_LINK =
   'https://developer.salesforce.com/tools/vscode/en/faq/telemetry';
 export const TELEMETRY_METADATA_COUNT = 'metadataCount';
 
+export const CONFIG_SET_EXECUTOR = 'force_config_set_org_text';
+export const CONFIG_SET_NAME = 'force_config_set_title';
+export const TABLE_NAME_COL = 'table_header_name';
+export const TABLE_VAL_COL = 'table_header_value';
+export const TABLE_SUCCESS_COL = 'table_header_success';
```
Comment (gbockus-sf) [inline packages/salesforcedx-vscode-core/src/constants.ts:32]: So close to alpha order...wanna swap the last two?

File: packages/salesforcedx-vscode-core/test/vscode-integration/commands/forceConfigSet.test.ts [hunk 12]
```diff
@@ -5,35 +5,69 @@
...
+    expect(formatOutput).to.contain(nls.localize(CONFIG_SET_NAME), DEFAULT_USERNAME_KEY);
+    expect(formatOutput).to.contain(usernameOrAlias, String(true));
+  });
+
+  it('should display error message in output channel', async () => {
+    setDefaultUsernameOrAliasStub.throws(new Error(errorMessage));
+    await forceConfigSet(usernameOrAlias);
+    expect(channelSpy.callCount).to.equal(2);
+    expect(channelSpy.lastCall.args.length).to.equal(1);
+    expect(channelSpy.lastCall.args[0]).to.contain(errorMessage);
```
Comment (gbockus-sf) [inline packages/salesforcedx-vscode-core/test/vscode-integration/commands/forceConfigSet.test.ts:71]: nice tests 👍

**Review by gbockus-sf — COMMENTED** [review]
Great stuff...Thanks for picking this up.  A few suggestions.

File: packages/salesforcedx-utils-vscode/test/jest/config/configUtil.test.ts (hunk 9 above)
Comment (RitamAgrawal) [reply packages/salesforcedx-utils-vscode/test/jest/config/configUtil.test.ts:44]: No it's not! I need to update the name.

File: packages/salesforcedx-vscode-core/src/constants.ts (hunk 11 above)
Comment (RitamAgrawal) [reply packages/salesforcedx-vscode-core/src/constants.ts:32]: Thanks for pointing this one out.

File: packages/salesforcedx-utils-vscode/src/config/configUtil.ts (hunk 8 above)
Comment (gbockus-sf) [reply packages/salesforcedx-utils-vscode/src/config/configUtil.ts:157]: Ritam and I dug into this together and found that this isn't validating the username, but actually verifying there is a valid connection.  She's going to update the comment to more clearly state what is being checked and leave the function as is.

Comment (RitamAgrawal) [reply packages/salesforcedx-utils-vscode/src/config/configUtil.ts:157]: `Org.create()`creates a connection with the active usernameOrAlias and throws an error if the usernameOrAlias is expired. Will update the comment for it to make more sense.

File: packages/salesforcedx-vscode-core/src/constants.ts (hunk 11 above)
Comment (RitamAgrawal) [reply packages/salesforcedx-vscode-core/src/constants.ts:32]: Checked it by swapping the order! It throws a linting error.

File: packages/salesforcedx-vscode-core/test/vscode-integration/commands/auth/forceAuthLogout.test.ts [hunk 13]
```diff
@@ -191,10 +191,10 @@ describe('Force Auth Logout Default', () => {
     expect(
       notificationStub.callCount,
       'should not have posted an error message'
-    ).to.equal(1);
+    ).to.equal(2);
     const notificationArgs = notificationStub.getCall(0).args;
     expect(notificationArgs).to.deep.equal([
-      'SFDX: Log Out from Default Org successfully ran',
+      'SFDX: Set a Default Org successfully ran',
       'Show',
```
Comment (RitamAgrawal) [inline packages/salesforcedx-vscode-core/test/vscode-integration/commands/auth/forceAuthLogout.test.ts:198]: I have updated the tests for now and logged a new WI [W-12431853](https://gus.lightning.force.com/lightning/r/ADM_Work__c/a07EE00001K0MdvYAF/view) to address the two notification pop-ups while running `SFDX: Log Out from Default Org` to log out from a scratch org.

**Review by randi274 — APPROVED** [review]
QE Approved:
✅ Change default org w/ one Alias -> shouldn't be using CLI, should display as formatted table in the output channel
✅ Change default org w/ two Aliases -> should use the first one as the Value
✅ Change default org w/ no Alias -> shows the username as the Value
✅ Confirmed the same on Windows

**Pull request merged**